src/luarocks/fetch/git.lua

```diff
@@ -86,7 +92,20 @@ function git.get_sources(rockspec, extract, dest_dir, depth)
    local ok, err = fs.change_dir(store_dir)
    if not ok then return nil, err end
 
-   local command = {fs.Q(git_cmd), "clone", depth or "--depth=1", rockspec.source.url, module}
+   -- shallow clone by default
+   local depth = "--depth=1"
+   if rockspec.source.revision then
+      if rockspec.source.branch and git_supports_single_branch(git_cmd) then
+         -- If the rockspec specifies a branch, then clone only that branch
+         depth = "--single-branch"
+      else
+         -- otherwise, we need a full clone to make sure we fetch all available
+         -- revisions
+         depth = "--"
```

Comment:
This seems prone to later issues: `--` signifies "things after this point are arguments not options".